apps/labs/posts/scikit-lego-narwhals.md

```diff
@@ -0,0 +1,101 @@
+category: [PyData ecosystem]
+---
+
+# How scikit-lego became dataframe-agnostic
+
+Scikit-lego is a Python project which provides extra building blocks on top of scikit-learn.
+It provides some extra funcionality such as `TimeGapSplit` and `add_lags` which may be considered
+too niche for scikit-learn itself. These functions have always only ever supported pandas...up until now!
+
+As of scikit-lego 0.9.0, you can also pass dataframes from Polars, Modin, cuDF, and in principle a whole host
+of other dataframes.
+But, how does it work? Why did they use Narwhals?
+Why don't they just
+convert to pandas internally on the user's behalf?
+
+We'll start by answering these questions, and will end with some hopes for the future.
+
+## How does it work?
+
+Let's take a look at `sklego.pandas_utils.add_lags`. The code before version 0.9.0 did something like this:
+
+- check if the input is NumPy or a pandas dataframe
+- if it's NumPy, then perform array operations
+- if it's pandas, then perform pandas dataframe operations
+
+In order to support Polars too, one possible solution could have been do add a third step to
+the above:
+
+- if it's Polars, then perform Polars dataframe operations
+
+This looks simple, but it actually opens up a rabbit-hole of extra steps, such as:
+
+- if it's Modin, ...
+- if it's cuDF, ...
+- if it's fireducks, ...
+
+In order to address all the above, scikit-lego opted for a simpler approach: use Narwhals.
+The steps then become:
+
+- run `narwhals.from_native` on the input
+- express the dataframe logic using the subset of the Polars API supported by Narwhals
+
+Like this, Narwhals takes care of translating its (Polars-like) syntax to each
+supported backend. Read the [Narwhals docs](https://narwhals-dev.github.io/narwhals) for more
+info on how to use it, and for a tutorial.
+
+## Why not "just convert to pandas"?
+
+There are different ways of converting non-pandas input to pandas. Rather than using Narwhals,
+couldn't scikit-lego just have done something like:
+
+- if the input is pandas, keep as-is
+- if the input is a non-pandas dataframe, then convert to pandas
+
+However, even in the best-case zero-copy-conversion scenario, this presents several issues:
+
+- once the input has been converted to pandas, there's no standardised way of converting back to
+  the original dataframe library.
+- using pandas as an engine may be significantly slower than doing everything using the original
+  dataframe
+
+Furthermore, converting to pandas may present a cost - for example, if you start with a Polars
+LazyFrame, then you're required to call `.collect` on it before converting
+to pandas.
+
+To drive the second point home, here are some timings comparing running `add_lags` on a
+10-million-row Polars dataframe directly, as opposed to converting to pandas and back:
+
+- passing Polars directly to `add_lags`:  ~ 10.8 ms
+- converting Polars to pandas, passing that to `add_lags`, then converting back: ~ 1.47 s
+
```

Comment:
Often when comparing two results, you use the same units to make it easier for the reader to compare.  Writing 1470ms would make it clear that it's almost 150x faster.

(Also, SI says no space between numbers and units, but I gave up that battle a long time ago, as long as the paper is consistent, which this is. :-) )